### ISSUE TITLE
Python crashes when calling sys.exit()

### ISSUE THREAD

**Issue by Dade916** [LuxCoreRender/LuxCore#29]
Python crashes when calling sys.exit() if pyluxcore.Init() was called with an call back. This is due to pyluxcore keeping a reference to the callback in a static boost::python::object. Everything works fine if python code exit normally without calling sys.exit().

I haven't yet found a work around to this problem.

**Comment by Dade916** [comment]
I have added a new SetLogHandler() function to the LuxCore API so it is now possible a workaround:

```
try:
	pyluxcore.Init(loghandler.LuxCoreLogHandler)

	cmd.main()
finally:
	pyluxcore.SetLogHandler(None)
```

Calling "pyluxcore.SetLogHandler(None)" before the end fix the problem.

**Comment by tschw** [comment]
So the destructor of the static object attempts to clean up but the interpreter's gone?

There's [atexit](https://docs.python.org/3/library/atexit.html) in Python (similar to [std::atexit](http://en.cppreference.com/w/cpp/utility/program/atexit) in C++) that lets you register a callback that is invoked before the interpreter dies. That might do to perform a safe tear-down...

**Comment by Dade916** [comment]
Yes there is a static object holding the reference to a python object (i.e. a function call back to print log messages) and when the destructor of the static object is called, the python interpreter as already gone.

It can be solved with the try/finally or with the atexit but both solutions are not transparent for the API user: they must be coded in the application.

**Comment by tschw** [comment]
Yes, I meant doing the python `atexit` call programmatically from within C++ code, actually.

Another solution would be to have everything, including the system context in classes, then it would get destroyed when the interpreter shuts down and the cleanup could happen there.
